internal/xpkg/layers.go

```diff
@@ -156,3 +159,106 @@ func AnnotateLayers(i v1.Image) (v1.Image, error) {
 
 	return mutate.ConfigFile(img, cfgFile)
 }
+
+// ImageFromFiles creates a v1.Image from arbitrary files on disk.
+// Each top-level directory at `root` is a separate layer.
+// The function performs no interpretation (parsing) of the files.
+func ImageFromFiles(root string) (v1.Image, error) {
+	extManifest := empty.Image
+
+	entries, err := os.ReadDir(root)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		curDir := filepath.Join(root, entry.Name())
+
+		// Since there is an arbitrary directory of mostly small files, we'll
+		// forego streaming in-memory at the expense of some disk I/O with temporary tarballs.
+		tmpFile, err := os.CreateTemp("", "extension-*.tar")
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = tmpFile.Close() }()
+		if err := createTarball(curDir, tmpFile.Name()); err != nil {
+			return nil, err
+		}
+
+		// Create layer from the tarball file
+		layer, err := tarball.LayerFromFile(tmpFile.Name())
+		if err != nil {
+			return nil, err
+		}
+		// Append layer from the dir tarball
+		extManifest, err = mutate.Append(
+			extManifest,
+			mutate.Addendum{
+				Layer: layer,
+				Annotations: map[string]string{
+					AnnotationKey: entry.Name(),
+				},
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return extManifest, nil
+}
+
+func createTarball(in string, out string) error {
+	f, err := os.Create(filepath.Clean(out))
+	if err != nil {
+		return err
+	}
+	defer func() { _ = f.Close() }()
+
+	gw := gzip.NewWriter(f)
+	defer func() { _ = gw.Close() }()
+
+	tw := tar.NewWriter(gw)
+	defer func() { _ = tw.Close() }()
+
+	return filepath.Walk(in, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Get the relative path
+		relPath, err := filepath.Rel(filepath.Dir(in), path)
+		if err != nil {
+			return err
+		}
+
+		// Create tar header
+		header, err := tar.FileInfoHeader(info, info.Name())
+		if err != nil {
+			return err
+		}
+		header.Name = relPath
+
+		if err := tw.WriteHeader(header); err != nil {
+			return err
+		}
+
+		// If not a directory, write file content
+		if !info.IsDir() {
+			f, err := os.Open(filepath.Clean(path))
+			if err != nil {
+				return err
+			}
+			defer func() { _ = f.Close() }()
+
+			if _, err := io.Copy(tw, f); err != nil {
+				return err
```

Comment:
Will these errors provide the user enough context to figure out what broke? e.g. File paths etc. If not, it'd be a good idea to `errors.Wrap` them with that information.

Reply:
Also in the past io.Copy` has caused a security linter error. Wonder why it isn't here.

https://github.com/crossplane/function-runtime-oci/blob/3d2dcd121ffc811a52636b357ef0531a189e392d/internal/oci/store/store.go#L340-L358

Reply:
> Also in the past io.Copy` has caused a security linter error. Wonder why it isn't here.

I believe `gosec` may not flag it if the source io.Reader (the temp file in this case) isn't influenced by user input. I don't know the specifics about how it determines that, though.

Reply:
AFAICT io.Copy errors are fairly obvious and tell you either something went wrong reading (unexpected EOF) or something went wrong writing (short buffer, disk full)

Reply:
I updated this to use `errors.Wrap` as suggested.